iota-trinary/src/trytes.rs

```diff
@@ -0,0 +1,675 @@
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use iota_trinary::*;
+    ///
+    /// let mut t = Trytes::try_from("ABC").unwrap();
+    ///
+    /// assert_eq!(t.len(), 3);
+    /// ```
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Returns `true` if this `Trytes` has a length of zero, and `false` otherwise.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use iota_trinary::*;
+    ///
+    /// let mut t = Trytes::new();
+    /// assert!(t.is_empty());
+    ///
+    /// t.push('9');
+    /// assert!(!t.is_empty());
+    /// ```
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Splits the trytes into two at the given index.
+    ///
+    /// Returns a newly allocated `Trytes`. `self` contains bytes `[0, at)`, and
+    /// the returned `Trytes` contains bytes `[at, len)`.
+    ///
+    /// Note that the capacity of `self` does not change.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `at` is beyond the last code point of the trytes.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use iota_trinary::*;
+    ///
+    /// let mut t = Trytes::try_from("IOTA").unwrap();
+    /// let t2 = t.split_off(2);
+    /// assert_eq!(t.as_str(), "IO");
+    /// assert_eq!(t2.as_str(), "TA");
+    /// ```
+    pub fn split_off(&mut self, at: usize) -> Self {
+        let other = self.0.split_off(at);
+        Trytes(other)
+    }
+
+    /// Truncates this `Trytes`, removing all contents.
+    ///
+    /// While this means the `Trytes` will have a length of zero, it does not
+    /// touch its capacity.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use iota_trinary::*;
+    ///
+    /// let mut t = Trytes::try_from("FOO").unwrap();
+    ///
+    /// t.clear();
+    ///
+    /// assert!(t.is_empty());
+    /// assert_eq!(0, t.len());
+    /// ```
+    pub fn clear(&mut self) {
+        self.0.clear()
+    }
+
+    /// Encode a string literal as bytes to trytes.
+    ///
+    /// ```
+    /// use iota_trinary::*;
+    ///
+    /// let mut t = Trytes::new();
+    ///
+    /// t.encode("Lorem");
+    ///
+    /// assert_eq!("VBCDFDTCAD", t.as_str());
+    /// ```
+    pub fn encode(&mut self, plain: &str) {
+        for byte in plain.bytes() {
+            let first = byte % 27;
+            let second = (byte - first) / 27;
+            self.0.put(TRYTE_ALPHABET[first as usize]);
+            self.0.put(TRYTE_ALPHABET[second as usize]);
+        }
+    }
+
+    /// ```
+    /// use iota_trinary::*;
+    ///
+    /// let t = Trytes::try_from("VBCDFDTCAD").unwrap();
+    ///
+    /// assert_eq!("Lorem", t.decode());
+    /// ```
+    pub fn decode(&self) -> String {
+        let mut s = String::with_capacity(self.len() / 2);
+        for slice in self.chunks(2) {
+            let first: u8 = TRYTE_ALPHABET.iter().position(|&x| x == slice[0]).unwrap() as u8;
+            let second: u8 = TRYTE_ALPHABET.iter().position(|&x| x == slice[1]).unwrap() as u8;
```

Comment:
Instead of walking the entire array here, it would be more efficient to create a function that goes the other way. Something like this:

```rust
fn chat_to_tryte(c: char) -> Option<Tryte> {
    match c {
        'A'..'B' => Some(c as u8 - b'A'),
        '9' => Some(26),
        _ => None,
    }
}
```